src/Database/LSMTree/Internal/PageAcc.hs

```diff
@@ -225,8 +236,8 @@ pageAccAddElem PageAcc {..} k e
                + (case n of { 0 -> 6; 1 -> 2; _ -> 4 })  -- key and value offsets
                + sizeofEntry k e
 
-        -- check for size overflow
-        if s' > pageSize
+        if s' > pageSize || -- check for size overflow
+           n >= maxKeys     -- check for buffer overflow
```

Comment:
Is this for extra safety, so we can never exceed the hardcoded max number of keys, or is absolutely required for correctness?

Reply:
Ah, I think this part of the commit comment explains it:

> To be safe (i.e. avoid indexing out of bounds) even in the case of all
empty keys, adjust the overflow test to also check the number of keys.
This means it fails to produce a result for such (extreme) test cases,
which is a difference versus the reference implementation, but this
difference can be handled in the tests explicitly without much
difficulty.